spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/WhitelistedSerializationStrategy.java

```diff
@@ -0,0 +1,162 @@
+            oos.writeObject(state);
+            oos.flush();
+            return bos.toByteArray();
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e);
+        } finally {
+            if (oos != null) {
+                try {
+                    oos.close();
+                } catch (IOException e) {
+                    // eat it
+                }
+            }
+        }
+    }
+
+    public <T> T deserialize(byte[] byteArray) {
+        ObjectInputStream oip = null;
+        try {
+            oip = new WhitelistedObjectInputStream(new ByteArrayInputStream(byteArray),
+                    Thread.currentThread().getContextClassLoader(), allowedClasses);
+            @SuppressWarnings("unchecked")
+            T result = (T) oip.readObject();
+            return result;
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e);
+        } catch (ClassNotFoundException e) {
+            throw new IllegalArgumentException(e);
+        } finally {
+            if (oip != null) {
+                try {
+                    oip.close();
+                } catch (IOException e) {
+                    // eat it
+                }
+            }
+        }
+    }
+
+    /**
+     * Special ObjectInputStream subclass that checks if classes are allowed to deserialize. The class
+     * should be configured with a whitelist of only allowed (safe) classes to deserialize.
+     */
+    private static class WhitelistedObjectInputStream extends ObjectInputStream {
+
+        /**
+         * The list of classes which are allowed for deserialization.
+         */
+        private final List<String> allowedClasses;
+
+        /**
+         * The class loader to use for loading local classes.
+         */
+        private final ClassLoader classLoader;
+
+        /**
+         * Create a new WhitelistedObjectInputStream for the given InputStream, class loader and
+         * allowed class names.
+         *
+         * @param in             The InputStream to read from.
+         * @param classLoader    The ClassLoader to use for loading local classes.
+         * @param allowedClasses The list of allowed classes for deserialization.
+         * @throws IOException If something went wrong.
+         */
+        private WhitelistedObjectInputStream(InputStream in, ClassLoader classLoader, List<String> allowedClasses)
+                throws IOException {
+            super(in);
+            this.classLoader = classLoader;
+            this.allowedClasses = Collections.unmodifiableList(allowedClasses);
+        }
+
+        /**
+         * Resolve the class only if it's allowed to deserialize.
+         *
+         * @see ObjectInputStream#resolveClass(ObjectStreamClass)
+         */
+        @Override
+        protected Class<?> resolveClass(ObjectStreamClass classDesc)
+                throws IOException, ClassNotFoundException {
+            if (isProhibited(classDesc.getName())) {
```

Comment:
As it stands now, when an application upgrades to 2.4 and have extended `OAuth2AccessToken` (or `OAuth2RefreshToken` and `OAuth2Authentication`), it will be forced to configure a custom `WhitelistedSerializationStrategy`, otherwise the application will fail when attempting to deserialize.

The preference would be that the application will continue to work after upgrading to 2.4 without having to provide an additional configuration (code change). Can we add an additional check for superclass? For example, `classDesc.forClass().getSuperclass()`

Reply:
@jgrandja 

> As it stands now, when an application upgrades to 2.4 and have extended `OAuth2AccessToken` (or `OAuth2RefreshToken`, `OAuth2Authentication`), it will be forced to configure a custom `WhitelistedSerializationStrategy`, otherwise the application will fail when attempting to deserialize.

Unfortunately yes, that's how it's going to be.

> The preference would be that the application will continue to work after upgrading to 2.4 without having to provide an additional configuration (code change). 

Yes, ideally.

> Can we add an additional check for superclass? For example, `classDesc.forClass().getSuperclass()`

I am not sure if we should do that. If we add such a check, an attacker can prepare a class `MaliciousOAuth2AccessToken` which extends `DefaultOAuth2AccessToken`, and put a malicious code to the constructor. Then, `MaliciousOAuth2AccessToken` can pass the `is Prohibited()` check.

Reply:
@artem-smotrakov 

> I am not sure if we should do that. If we add such a check, an attacker can prepare a class `MaliciousOAuth2AccessToken`...

Yes, you're right. Ok, so we should not apply check on superclass?

> As it stands now, when an application upgrades to 2.4 and have extended `OAuth2AccessToken` (or `OAuth2RefreshToken`, `OAuth2Authentication`), it will be forced to configure a custom `WhitelistedSerializationStrategy`, otherwise the application will fail when attempting to deserialize.

This change introduces incompatibility changes between 2.3.x and 2.4.0, which cannot happen in a minor  release. An application built against 2.3.x should be able to upgrade to 2.4.0 without any issues. This change would be fine for a major version though. Please see [Versioning](https://apr.apache.org/versioning.html#strategy) for additional details.

In order to get this change into 2.4.0, this will need to be an "opt-in" feature rather than the default configuration. So if the application wants stricter deserialization, than they would need to explicitly configure `WhitelistedSerializationStrategy`. For the default configuration, the application should perform the same as it did in 2.3.x.

Reply:
@jgrandja 

> In order to get this change into 2.4.0, this will need to be an "opt-in" feature rather than the default configuration.

That's sad. I would prefer to make it a bit more secure by default. I'll update it to be an opt-it feature, however I don't like it.

Reply:
@artem-smotrakov 

> I would prefer to make it a bit more secure by default

I definitely would too. But with the way versioning works, this would break existing applications and users would log new issues around this as we have seen with the 2.3.7 release.

Reply:
@jgrandja 

* I added a new `DefaultSerializationStrategy` which allows all classes by default. The `SerializationUtils` uses `DefaultSerializationStrategy` by default.
* I also mentioned in javadoc that the default settings may be unsafe if data come from an untrusted source.
* Updated tests

Please let me know what you think.